### PR TITLE
Module 3

### PR DESCRIPTION
**Pull request by cpujades**
@gsantam @guillerbf 

I have completed the module 3 assessment in the file "module_3.md / module_3.ipynb", the different sections are marked with markdown.

I have several doubts when choosing the right threshold for the linear model and executing the pipeline.

### CODE REVIEW

File: src/module_3/module_3.md [hunk 1]
```diff
@@ -0,0 +1,891 @@
+```python
+import pandas as pd
+import numpy as np
+import matplotlib.pyplot as plt
+import seaborn as sns
+```
+
+
+```python
+df = pd.read_csv('/Users/carlospujades/Documents/Carletes/Myself/Zrive/Module 2/feature_frame.csv')
```
Comment (gsantam) [inline src/module_3/module_3.md:10]: A short explanation of the df (what are the rows, what are the columns...) would be great to refresh our memory

File: src/module_3/module_3.md [hunk 2]
```diff
@@ -0,0 +1,891 @@
...
+
+
+
+### Filtering for orders with 5 or more items
+
+
+```python
+items_per_order = df.groupby('order_id')['outcome'].sum()
+big_orders = items_per_order[items_per_order >= 5]
+orders_nums = big_orders.index
```
Comment (gsantam) [inline src/module_3/module_3.md:192]: as a rule of thumb, I'd never use `index`. Instead use a unique identifier like `order_id`

File: src/module_3/module_3.md [hunk 3]
```diff
@@ -0,0 +1,891 @@
...
+     24  days_since_purchase_product_type  float64
+     25  avg_days_to_buy_product_type      float64
+     26  std_days_to_buy_product_type      float64
+    dtypes: float64(19), int64(4), object(4)
+    memory usage: 462.3+ MB
+
+
+
+```python
+df.drop(['order_id', 'created_at', 'order_date'], inplace=True, axis=1)
```
Comment (gsantam) [inline src/module_3/module_3.md:241]: I wouldn't drop the `order_id` identifier

File: src/module_3/module_3.md [hunk 4]
```diff
@@ -0,0 +1,891 @@
...
+
+columns = ['user_order_seq', 'outcome', 'ordered_before', 'abandoned_before', 'active_snoozed', 'set_as_regular', 'count_adults', 'count_children', 'count_babies', 'count_pets', 'people_ex_baby', 'days_since_purchase_variant_id', 'days_since_purchase_product_type']
+for c in columns:
+    df[c] = df[c].astype('int32')
```
Comment (gsantam) [inline src/module_3/module_3.md:245]: good idea, but I would make sure that this is not screwing your numbers since can be a little bit dangerous

File: src/module_3/module_3.md [hunk 5]
```diff
@@ -0,0 +1,891 @@
...
+    Name: count, dtype: int64
+
+
+
+
+```python
+X = df.drop('outcome', axis=1)
+y = df['outcome']
+
+X_train, X_test, y_train, y_test = train_test_split(X, y, stratify=y, test_size=0.4, random_state=9)
```
Comment (gsantam) [inline src/module_3/module_3.md:495]: stratify here isn't gonna make a big difference tbh, it's more used when you have a lot of categoriues, some of them very underrepresented

File: src/module_3/module_3.md [hunk 6]
```diff
@@ -0,0 +1,891 @@
...
+```
+
+
+```python
+cat_cols = ['product_type', 'vendor']
+for c in cat_cols:
+    train_freq = X_train[c].value_counts()
+    X_train[c] = X_train[c].map(train_freq)
+    
+    test_freq = X_test[c].value_counts()
```
Comment (gsantam) [inline src/module_3/module_3.md:505]: What happens if you have a product_type/vendor which is in test but not in train? The model would probably fail to predict that

File: src/module_3/module_3.md [hunk 7]
```diff
@@ -0,0 +1,891 @@
...
+
+
+```python
+from sklearn.model_selection import GridSearchCV
+from sklearn.metrics import classification_report
+from sklearn.model_selection import cross_validate
+from imblearn.over_sampling import SMOTE
+
+
+sgd_log = SGDClassifier(loss='log_loss',class_weight='balanced')
```
Comment (gsantam) [inline src/module_3/module_3.md:525]: I wouldn't use `balance` unless you have a very big class imbalance (which is not your case): it doesn't really change anything in terms of performance (try it yourself) and it distorts the probabilities produced by the model

File: src/module_3/module_3.md [hunk 8]
```diff
@@ -0,0 +1,891 @@
...
+sgd_svm = SGDClassifier(loss='hinge', class_weight='balanced')
+ridge = RidgeClassifier(class_weight='balanced')
+log_reg = LogisticRegression(class_weight='balanced')
+
+oversample = SMOTE()
```
Comment (gsantam) [inline src/module_3/module_3.md:530]: I wouldn't use smote since it's not useful

File: src/module_3/module_3.md [hunk 9]
```diff
@@ -0,0 +1,891 @@
...
+      <td>0.800841</td>
+      <td>0.638327</td>
+    </tr>
+  </tbody>
+</table>
+</div>
+
+
+
+</p>As whe have class imbalance, we use `class_weight='balance'` and oversample to increase the number of samples in the underrepresented class.</p>
```
Comment (gsantam) [inline src/module_3/module_3.md:621]: As said above: your class imbalance is not very big, don't use these methods

File: src/module_3/module_3.md [hunk 10]
```diff
@@ -0,0 +1,891 @@
...
+<p>Based on the previous cells, we stick with Logistic Regression that has a higher score across accuracy, precision and recall. Although, due to the class imbalance the metric we are most interested in is recall.</p>
```
Comment (gsantam) [inline src/module_3/module_3.md:622]: At what threshold is this precision / recall? Why are you using this threshold and why it's important for the business? Could you use a PR curve instead to inform to the business of what would happen with different thresholds?

File: src/module_3/module_3.md [hunk 11]
```diff
@@ -0,0 +1,891 @@
...
+y_pred = grid.predict(X_test_scaled)
+y_probs = grid.predict_proba(X_test_scaled)[:,1]
+y_log_probs = grid.predict_log_proba(X_test_scaled)[:,1]
+print(classification_report(y_test, y_pred))
+print(confusion_matrix(y_test, y_pred))
+```
+
+    {'C': 0.001}
+    0.800954386060868
+    LogisticRegression(C=0.001, class_weight='balanced')
```
Comment (gsantam) [inline src/module_3/module_3.md:651]: Again, you are using a concrete threshold when probably the business wants to choose a concrete tradeoff of precision / recall among all of them

File: src/module_3/module_3.md [hunk 12]
```diff
@@ -0,0 +1,891 @@
...
+print("Recall:", recall_score(y_test, y_pred_new_threshold)) 
+print("F1 score:", f1_score(y_test, y_pred_new_threshold))
+print(confusion_matrix(y_test, y_pred_new_threshold))
+```
+
+    Evaluation metrics with threshold 0.914:
+    Accuracy: 0.9729430602762072
+    Precision: 0.18848706797894255
+    Recall: 0.26284711139482925
+    F1 score: 0.21954145561183686
```
Comment (gsantam) [inline src/module_3/module_3.md:793]: add PR AUC to the metrics

File: src/module_3/module_3.md (hunk 5 above)
Comment (gsantam) [reply src/module_3/module_3.md:495]: More important (actually probably the most important thing of this exercise): I think this train / test split doesn't work for different reasons. First of all you are putting products of the same order in train and test at the same time. This is wrong since you are leaking data from train to test (if a concrete product was bought for an order, the rest of products have probably higher chances of being bought, but you can't know this in runtime). 
Also, you are ignoring the time component: you are putting contemporary orders at the same time in train and test. 
In general, the best way of designing a train test strategy is to think of how are you going to use the model, and what data is gonna be available for training and for testing. If you think of this you'll realise that in runtime we are only going to have available these orders which happened in the PAST and we are gonna predict on orders happening in the FUTURE. This is not what your train / test reflect since you are putting orders happening at the same time either in train and test.

**Review by gsantam — CHANGES_REQUESTED** [review]
The main problem relies on the train / test schema which is wrong. I left a comment on it, but we can discuss if you need more details or help
